File: app/src/main/java/br/fmu/bioreino/adapter/ListaCategoriasAdapter.java
package br.fmu.bioreino.adapter;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import br.fmu.bioreino.R;
import br.fmu.bioreino.dao.CursosDAO;
import br.fmu.bioreino.model.Categoria;
import br.fmu.bioreino.util.CategoriaInterface;
import br.fmu.bioreino.util.ListaUtil;

public class ListaCategoriasAdapter extends RecyclerView.Adapter<ListaCategoriasAdapter.CategoriaLayout> implements Filterable {

    private final CategoriaInterface categoriaInterface;

    private final ArrayList<Categoria> categorias = new ArrayList<>();
    private final ArrayList<Categoria> categoriasCompleta = CursosDAO.getCategorias();

    public ListaCategoriasAdapter(CategoriaInterface categoriaInterface) {
        this.categoriaInterface = categoriaInterface;
    }


    // Métodos Padrão ------------------------------------------------------------------------------
    @NonNull
    @Override
    public CategoriaLayout onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new CategoriaLayout(
                ListaUtil.inflaLayout(parent, R.layout.item_categoria),
                categoriaInterface
        );
    }

    @Override
    public void onBindViewHolder(@NonNull CategoriaLayout holder, int position) {
        Categoria categoria = categorias.get(position);
        atualizaInformacoes(holder, categoria);
    }

    @Override
    public int getItemCount() {
        return categorias.size();
    }

    public static class CategoriaLayout extends RecyclerView.ViewHolder {

        final CardView cardFundo;
        final TextView texto;

        public CategoriaLayout(@NonNull View itemView, CategoriaInterface categoriaInterface) {
            super(itemView);

            cardFundo = itemView.findViewById(R.id.item_categoria_card);
            texto = itemView.findViewById(R.id.item_categoria_texto);

            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    if (categoriaInterface != null) {
                        categoriaInterface.quandoClicarNaCategoria(texto.getText().toString().toLowerCase(Locale.ROOT));
                    }
                }
            });
        }

    }


    // Filtro --------------------------------------------------------------------------------------
    @Override
    public Filter getFilter() {
        return filtro;
    }

    private final Filter filtro = new Filter() {
        @Override
        protected FilterResults performFiltering(CharSequence charSequence) {
            List<Categoria> listaFiltrada = new ArrayList<>();

            if (charSequence == null || charSequence.length() == 0) {
                listaFiltrada.addAll(categoriasCompleta);
            } else {
                String planoSelecionado = charSequence.toString().toLowerCase(Locale.ROOT).trim();

                for (Categoria categoria : categoriasCompleta) {
                    if (categoria.getPlano().toLowerCase(Locale.ROOT).contains(planoSelecionado)) {
                        listaFiltrada.add(categoria);
                    }
                }
            }

            FilterResults resultado = new FilterResults();
            resultado.values = listaFiltrada;

            return resultado;
        }

        @Override
        protected void publishResults(CharSequence charSequence, FilterResults filterResults) {
            categorias.clear();
            categorias.addAll((List) filterResults.values);
            notifyDataSetChanged();
        }

    };


    // Helpers -------------------------------------------------------------------------------------
    private void atualizaInformacoes(@NonNull CategoriaLayout holder, Categoria categoria) {

        holder.cardFundo.setCardBackgroundColor(categoria.getCorFundo());
        holder.texto.setText(categoria.getNome());
        holder.texto.setTextColor(categoria.getCorTexto());

    }

}

File: app/src/main/java/br/fmu/bioreino/viewModel/adapter/ListaCategoriasAdapter.kt
package br.fmu.bioreino.viewModel.adapter

import android.view.View
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.RecyclerView
import br.fmu.bioreino.R
import br.fmu.bioreino.viewModel.adapter.ListaCategoriasAdapter.CategoriaLayout
import br.fmu.bioreino.viewModel.util.CategoriaInterface
import br.fmu.bioreino.viewModel.util.ListaUtil.inflaLayout
import br.fmu.bioreino.model.dao.CursosDAO
import br.fmu.bioreino.model.entities.Categoria

class ListaCategoriasAdapter(private val categoriaInterface: CategoriaInterface) :
    RecyclerView.Adapter<CategoriaLayout>(), Filterable {
    private val categorias = ArrayList<Categoria>()
    private val categoriasCompleta = CursosDAO.categorias

    // Métodos Padrão ------------------------------------------------------------------------------
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CategoriaLayout {
        return CategoriaLayout(
            inflaLayout(parent, R.layout.item_categoria),
            categoriaInterface
        )
    }

    override fun onBindViewHolder(holder: CategoriaLayout, position: Int) {
        val categoria = categorias[position]
        atualizaInformacoes(holder, categoria)
    }

    override fun getItemCount(): Int {
        return categorias.size
    }

    class CategoriaLayout(itemView: View, categoriaInterface: CategoriaInterface?) :
        RecyclerView.ViewHolder(itemView) {
        val cardFundo: CardView
        val texto: TextView

        init {
            cardFundo = itemView.findViewById(R.id.item_categoria_card)
            texto = itemView.findViewById(R.id.item_categoria_texto)
            itemView.setOnClickListener {
                categoriaInterface?.quandoClicarNaCategoria(
                    texto.text.toString().lowercase()
                )
            }
        }
    }

    // Filtro --------------------------------------------------------------------------------------
    override fun getFilter(): Filter {
        return filtro
    }

    private val filtro: Filter = object : Filter() {
        override fun performFiltering(charSequence: CharSequence): FilterResults {
            val listaFiltrada: MutableList<Categoria> = ArrayList()
            if (charSequence.isEmpty()) {
                listaFiltrada.addAll(categoriasCompleta)
            } else {
                val planoSelecionado = charSequence.toString().lowercase().trim { it <= ' ' }
                for (categoria in categoriasCompleta) {
                    if (categoria.plano.lowercase().contains(planoSelecionado)) {
                        listaFiltrada.add(categoria)
                    }
                }
            }
            val resultado = FilterResults()
            resultado.values = listaFiltrada
            return resultado
        }

        override fun publishResults(charSequence: CharSequence, filterResults: FilterResults) {
            categorias.clear()
            categorias.addAll(filterResults.values as List<Categoria>)
            notifyDataSetChanged()
        }
    }

    // Helpers -------------------------------------------------------------------------------------
    private fun atualizaInformacoes(holder: CategoriaLayout, categoria: Categoria) {
        holder.cardFundo.setCardBackgroundColor(categoria.corFundo)
        holder.texto.text = categoria.nome
        holder.texto.setTextColor(categoria.corTexto)
    }
}